### PR TITLE
20180102 move to c++11 and ITKv5

### PR DESCRIPTION
**Pull request by hjmjohnson**
This huge Pull request is the first beta release of BRAINSTools that is going to start demanding use of C++11, and newer versions of CMake (3.8.2), and ITKv5 only support.

At this current stage, the code compiles, but it is dependant on other tool branches that are not yet updated to the master branches.

### CODE REVIEW

**Comment by hjmjohnson** [comment]
@Hui-Xie Please do a VERY careful review of each of these patches.  There are many changes happening to BRAINSTools, and you need to be aware of them all.

Hans

File: GTRACT/Cmdline/gtractCoRegAnatomy.cxx [hunk 1]
```diff
@@ -129,6 +130,17 @@ int main(int argc, char *argv[])
     return EXIT_FAILURE;
     }
 
+  std::string convertedVolume;
+  DWIConvert dwiConvert;
+  if (0 == dwiConvert.convertInputVolumeToNrrdOrNifti("Nrrd",
+                                                      inputVolume,convertedVolume)){
+    inputVolume = convertedVolume;
```
Comment (ihnorton) [inline GTRACT/Cmdline/gtractCoRegAnatomy.cxx]: ITK braces?

File: GTRACT/Cmdline/gtractConcatDwi.cxx [hunk 2]
```diff
@@ -88,6 +90,17 @@ int main(int argc, char *argv[])
     return EXIT_FAILURE;
     }
 
+  std::vector<std::string> inputVolumeNrrd;
+  DWIConvert dwiConvert;
+  if (0 == dwiConvert.convertInputVolumeVectorToNrrdOrNifti(dwiConvert.detectOuputVolumeType(outputVolume),
+                                                            inputVolume,inputVolumeNrrd)){
+    inputVolume = inputVolumeNrrd;
+  }
```
Comment (ihnorton) [inline GTRACT/Cmdline/gtractConcatDwi.cxx]: ITK braces?

File: BRAINSCommonLib/genericRegistrationHelper.hxx [hunk 3]
```diff
@@ -350,6 +350,7 @@ MultiModal3DMutualRegistrationHelper<TTransformType, TOptimizer, TFixedImage,
   m_Registration->SetMetricSamplingStrategy(
                 static_cast<typename RegistrationType::MetricSamplingStrategyType>( m_SamplingStrategy ));
   m_Registration->SetMetricSamplingPercentage(this->m_SamplingPercentage);
+  m_Registration->MetricSamplingReinitializeSeed(121212);
```
Comment (ihnorton) [inline BRAINSCommonLib/genericRegistrationHelper.hxx]: Maybe this should be `const` at the top of file?

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+not necessary to contribute to BRAINSTools.
+
+You may visit the *Pro Git: Setup* resource in [GitHelp] for further
+information on setting up your local Git environment.
+
+Workflow
+--------
+
+BRAINSTools development uses a strategy similar to ITK
+[branchy workflow]() based on topic
```
Comment (ihnorton) [inline .github/CONTRIBUTING.md]: missing link?

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+   $ git prepush
+```
+
+Push commits in your topic branch for review by the community:
+
+```sh
+   $ git push mywriteaccessfork  my-topic
+```
+
+ind your change in the [BRAINSTools Github page] instance and add reviewers, or provide additional documentation.
```
Comment (ihnorton) [inline .github/CONTRIBUTING.md]: (https://github.com/BRAINSia/BRAINSTools)

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+```sh
+   $ git checkout my-topic
+```
+
+Merge the topic, which is originally forked off the `master` branch, to
+`master` branch:
+
+```sh
+   $ git gerrit-merge
```
Comment (ihnorton) [inline .github/CONTRIBUTING.md]: Should be github, right?

File: BRAINSCommonLib/itkFindCenterOfBrainFilter.hxx [hunk 7]
```diff
@@ -271,7 +270,7 @@ FindCenterOfBrainFilter<TInputImage, TMaskImage>
                 }
               maskCountsInPlane[dIndex]++;
               }
-            // #ifdef USE_DEBUGGIN_IMAGES
+            // #ifdef USE_DEBUGGING_IMAGES
```
Comment (ihnorton) [inline BRAINSCommonLib/itkFindCenterOfBrainFilter.hxx]: no-op ifdef because this is commented out

File: BRAINSCut/cnpy/cnpy.cpp [hunk 8]
```diff
@@ -88,8 +88,8 @@ void cnpy::parse_npy_header(FILE* fp, unsigned int& word_size, unsigned int*& sh
     //byte order code | stands for not applicable.
```
Comment (ihnorton) [inline BRAINSCut/cnpy/cnpy.cpp]: Might want to pull in the upstream changes? They are using `reinterpret_cast` already, so the warnings should go away. https://github.com/rogersce/cnpy/blob/master/cnpy.cpp

**Pull request closed without merging**